### ISSUE TITLE
Path parameter confusion when relying on routing priority

### ISSUE THREAD

**Issue by ryanccn** [honojs/hono#1982]
### What version of Hono are you using?

3.12.3

### What runtime/platform is your app running on?

Cloudflare Workers

### What steps can reproduce the bug?

```typescript
import { Hono } from "hono";

const app = new Hono();

app.get("/prefix/:a", (c) => c.json({ a: c.req.param("a") }));
app.get("/:a", (c) => c.json({ a: c.req.param("a") }));
app.get("/:b/:a", (c) => c.json({ a: c.req.param("a"), b: c.req.param("b") }));
```

### What is the expected behavior?

Each of these routes should return proper responses according to the routing rules:

- `/prefix/hello` matches the first route and returns `{ a: "hello" }`
- `/hello` matches the second route and returns `{ a: "hello" }`
- `/another/hello` matches the third route and returns `{ a: "hello", b: "another" }`

### What do you see instead?

What the third route actually ends up returning is

```json
{
    "a": "another",
    "b": "another"
}
```

### Additional information

This bug does not occur without the `/prefix/:a` route.

**Comment by yusukebe** [comment]
@ryanccn 

Thanks for raising the issue. That's the bug of TrieRouter! I'll investigate it. Please wait a bit.

**Comment by EdamAme-x** [comment]
i create sandbox on stackblitz for this issue.
https://stackblitz.com/edit/stackblitz-starters-z45okh?file=index.js

**Comment by EdamAme-x** [comment]
I have confirmed the situation in this screenshot.  
[image](https://github.com/honojs/hono/assets/121654029/4d745ed0-b5e3-4f13-831a-c16d3f72eb47)

**Comment by yusukebe** [comment]
@ryanccn @EdamAme-x 

I've fixed this issue and released v3.12.4, including the change. Try it.

**Comment by ryanccn** [comment]
It's working now! Thanks for the timely fix
